File: Find the Safest Path in a Grid.py
class Solution:
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        n = len(grid)

        if grid[0][0] or grid[-1][-1]:
            return 0

        dp = [[2 * n] * n for _ in range(n)]

        q = deque([])
        for i in range(n):
            for j in range(n):
                if grid[i][j]:
                    q.append((0, i, j))

        while q:
            dist, i, j = q.popleft()
            if dp[i][j] != 2 * n:
                continue

            dp[i][j] = dist
            for x, y in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
                if 0 <= x < n and 0 <= y < n and dp[x][y] == 2 * n:
                    q.append((dist + 1, x, y))
        # print(dp)
        path = [(-dp[0][0], 0, 0)]
        walked = set()
        while path:
            # print(path)
            dist, i, j = heapq.heappop(path)

            if (i, j) in walked:
                continue

            if i == n - 1 and j == n - 1:
                return -dist

            walked.add((i, j))
            for x, y in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
                if 0 <= x < n and 0 <= y < n and (x, y) not in walked:
                    heapq.heappush(path, (max(dist, -dp[x][y]), x, y))

# class Solution:
#     def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
#         m = len(grid)
#         n = len(grid[0])
#
#         q = deque()
#         visited = set()
#
#         safe = [[-1] * m for _ in range(n)]
#         for i in range(m):
#             for j in range(n):
#                 if grid[i][j]:
#                     q.append((i, j, 0))
#                     visited.add((i, j))
#
#         while q:
#             i, j, val = q.popleft()
#             if safe[i][j] != -1:
#                 continue
#
#             safe[i][j] = val
#
#             for x,y in [(i+1,j), (i-1,j), (i, j+1), (i, j-1)]:
#                 if 0 <= x < m and 0 <= y < n and (x,y) not in visited:
#                     visited.add((x,y))
#                     q.append((x, y, val+1))
#
#
#         visited = {(0, 0)}
#         q = [(-safe[0][0], 0, 0)]
#
#         # print(safe)
#
#         while q:
#             val, i, j = heapq.heappop(q)
#
#             if i == m-1 and j == n-1:
#                 return -val
#
#             for x,y in [(i+1,j), (i-1,j), (i, j+1), (i, j-1)]:
#                 if 0 <= x < m and 0 <= y < n and (x,y) not in visited:
#                     visited.add((x,y))
#                     heapq.heappush(q, (max(val, -safe[x][y]), x, y))

File: Find the Safest Path in a Grid.cpp
class Solution {
public:
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        int n = grid.size();
        if (grid[0][0] || grid[n-1][n-1]) return 0;

        vector<vector<int>> dp;
        for (int i = 0; i < n; i++) {
            dp.emplace_back(n, 2*n);
        }

        deque<pair<int, pair<int, int>>> q;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j]) q.push_back(make_pair(0, make_pair(i, j)));
            }
        }

        while (!q.empty()) {
            auto kpair = q.front();
            q.pop_front();
            int dist = kpair.first, i = kpair.second.first, j = kpair.second.second;

            if (dp[i][j] != 2*n) continue;

            dp[i][j] = dist;
            vector<pair<int, int>> directions {
                {i-1, j},
                {i+1, j},
                {i, j-1},
                {i, j+1},
            };
            for (auto xy: directions) {
                int x = xy.first, y = xy.second;
                if (0 <= x && x < n && 0 <= y && y < n && dp[x][y] == 2*n) q.push_back(make_pair(dist+1, make_pair(x, y)));
            }

        }


        vector<tuple<int, int, int>> path{
            {dp[0][0], 0, 0}
        };
        unordered_set<int> walked;
        while (!path.empty()) {
            pop_heap(path.begin(), path.end());
            auto t = path.back();
            int dist = get<0>(t), i = get<1>(t), j = get<2>(t);
            // cout << "Dist: " <<  dist << " i: " << i << " j: " << j << endl;
            path.pop_back();

            int key = (i << 10) | j;
            if (walked.contains(key)) continue;
            walked.insert(key);

            if (i == n-1 && j == n-1) return dist;

            vector<pair<int, int>> directions {
                {i-1, j},
                {i+1, j},
                {i, j-1},
                {i, j+1},
            };

            for (auto xy: directions) {
                int x = xy.first, y = xy.second;
                if (0 <= x && x < n && 0 <= y && y < n && !walked.contains((x << 10) | y)) {
                    tuple<int, int, int> new_t{min(dist, dp[x][y]), x, y};
                    path.push_back(new_t);
                    push_heap(path.begin(), path.end());
                }
            }
        }
        return 0;
    }
};